lib/Conversion/GPUToROCM/ConvertKernelFuncToHSACO.cpp

```diff
@@ -0,0 +1,407 @@
+              });
+            })) {
+      module.emitError("Error linking ROCDL bitcode module from " + filename);
+      return failure();
+    }
+  }
+
+  return success();
+}
+
+// Returns whether the module uses any ROCDL bitcode functions.
+// This function may have false positives
+static bool couldNeedDeviceBitcode(const llvm::Module &llvmModule) {
+  for (const llvm::Function &llvmFunction : llvmModule.functions()) {
+    // This is a conservative approximation
+    //  - not all such functions are in ROCm-Device-Libs.
+    if (!llvmFunction.isIntrinsic() && llvmFunction.isDeclaration())
+      return true;
+  }
+  return false;
+}
+
+// Links ROCm-Device-Libs into the given module if the module needs it.
+static LogicalResult linkROCDLIfNecessary(llvm::Module &llvmModule,
+                                          ModuleOp module,
+                                          rocm::AMDGPUVersion amdgpuVersion,
+                                          llvm::StringRef rocdlDir) {
+
+  if (!couldNeedDeviceBitcode(llvmModule))
+    return success();
+
+  return linkWithBitcodeModules(llvmModule, module,
+                                getROCDLPaths(amdgpuVersion, rocdlDir));
+}
+
+// Emits the given module to HSA Code Object. targetMachine is an initialized
+// TargetMachine for the AMDGPU target.
+OwnedHSACO
+GpuKernelToHSACOPass::emitModuleToHSACO(llvm::Module &llvmModule,
+                                        ModuleOp module,
+                                        llvm::TargetMachine &targetMachine) {
+  llvm::SmallString<128> tempdirName;
+  if (llvm::sys::fs::createUniqueDirectory("/tmp/amdgpu_mlir", tempdirName)) {
+    module.emitError("Failed to create tempdir for generating HSACO\n");
+    return {};
+  }
+
+  std::error_code ec;
+  if (clDumpLLVMIR) {
+    // dump the LLVM IR to file...this is just for debugging purposes
+    llvm::Twine irFilename =
+        llvm::Twine(llvmModule.getModuleIdentifier()) + ".ll";
+    llvm::SmallString<128> irPath;
+    llvm::sys::path::append(irPath, tempdirName, irFilename);
+
+    llvm::raw_fd_ostream irFileStream(irPath, ec, llvm::sys::fs::F_None);
+    llvmModule.print(irFileStream, nullptr);
+    irFileStream.flush();
+  }
+
+  // dump the GCN ISA binary file
+  llvm::Twine isabinFilename =
+      llvm::Twine(llvmModule.getModuleIdentifier()) + ".o";
+  llvm::SmallString<128> isabinPath;
+  llvm::sys::path::append(isabinPath, tempdirName, isabinFilename);
+
+  llvm::legacy::PassManager codegenPasses;
+  llvm::SmallVector<char, 0> stream;
+  llvm::raw_svector_ostream pstream(stream);
+  llvm::raw_fd_ostream isabinFileStream(isabinPath, ec, llvm::sys::fs::F_Text);
+  llvmModule.setDataLayout(targetMachine.createDataLayout());
+  targetMachine.addPassesToEmitFile(codegenPasses, isabinFileStream, nullptr,
+                                    llvm::TargetMachine::CGFT_ObjectFile);
+  codegenPasses.run(llvmModule);
+  isabinFileStream.flush();
+
```

Comment:
I would try to keep maximum coverage for the test, since the pass is only available when the AMDGPU backend is built in, we should be able to get until there for the basic test. We could just reload the binary file `isabinFilename` print it as hexadecimal and return.

Something like:

```
if (testMode) {
  auto hsacoFileOrError = llvm::MemoryBuffer::getFileAsStream(isabinFilename);
  if ((ec = hsacoFileOrError.getError()))
    return {};
  std::unique_ptr<llvm::MemoryBuffer> hsacoFile = std::move(hsacoFileOrError.get());
  SmallVector<char, 1> buffer;
  {
    llvm::raw_svector_ostream os(buffer);
    for (auto char : hascoFile.getBuffer())
      os.write_hex(char);
  }
  return std::make_unique<std::vector<char>>(buffer.begin(), buffer.end());
}
```


The test would be able to check that we actually correctly generate an ELF binary (by checking the ELF magic in hexadecimal) and cover everything but the linker when ROCM is not available.